vcd/resource_vcd_service_account.go

```diff
@@ -0,0 +1,344 @@
+			},
+			"uri": {
+				Type:        schema.TypeString,
+				Optional:    true,
+				Description: "URI of the client using the service account, can be freely defined by the user",
+			},
+			"active": {
+				Type:        schema.TypeBool,
+				Optional:    true,
+				Description: "Status of the service account.",
+			},
+			"file_name": {
+				Type:        schema.TypeString,
+				Optional:    true,
+				Description: "Name of the file that the API token will be saved to",
+			},
+			"allow_token_file": {
+				Type:     schema.TypeBool,
+				Optional: true,
+				Default:  false,
+				Description: "Set this to true if you understand the security risks of using" +
+					" API token files and would like to suppress the warnings",
+			},
+		},
+	}
+}
+
+func resourceVcdServiceAccountCreate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	vcdClient := meta.(*VCDClient)
+
+	org, err := vcdClient.GetOrgFromResource(d)
+	if err != nil {
+		return diag.Errorf("[Service Account create] error getting Org from resource: %s", err)
+	}
+
+	adminOrg, err := vcdClient.GetAdminOrgFromResource(d)
+	if err != nil {
+		return diag.Errorf("[Service Account create] error getting AdminOrg from resource: %s", err)
+	}
+
+	saName := d.Get("name").(string)
+	softwareId := d.Get("software_id").(string)
+	softwareVersion := d.Get("software_version").(string)
+	uri := d.Get("uri").(string)
+	filename := d.Get("file_name").(string)
+
+	// Role needs to be sent in URN format, and the role name needs to be percent-encoded
+	// e.g urn:vcloud:role:Organization%20Administrator
+	roleId := d.Get("role_id").(string)
+	role, err := adminOrg.GetRoleById(roleId)
+	if err != nil {
+		return diag.Errorf("[Service Account create] error getting Role: %s", err)
+	}
+	escapedRoleName := url.PathEscape(role.Role.Name)
+	formattedRole := "urn:vcloud:role:" + escapedRoleName
+
+	active := d.Get("active").(bool)
+	allowTokenFile := d.Get("allow_token_file").(bool)
+
+	if active && !allowTokenFile {
+		return diag.Diagnostics{
+			diag.Diagnostic{
+				Severity: diag.Error,
+				Summary:  "The file " + filename + " should be considered sensitive information.",
+				Detail: "The file " + filename + " will contain the initial service account API " +
+					"and should be considered SENSITIVE INFORMATION. " +
+					"The token will be invalidated at the next use but a new one will be provided in the request's response. If you acknowledge this, add\n\n	allow_token_file = true\n\nto the provider settings.",
+			},
+		}
+	}
+
+	sa, err := vcdClient.CreateServiceAccount(org.Org.Name, saName, formattedRole, softwareId, softwareVersion, uri)
+	if err != nil {
+		return diag.Errorf("[Service Account create] error creating Service Account: %s", err)
+	}
+	d.SetId(sa.ServiceAccount.ID)
+
+	useragent := vcdClient.Client.UserAgent
+	err = updateServiceAccountStatus(sa, active, filename, useragent)
```

Comment:
I think this part should remain abstracted in the go-vcloud-director SDK - it should handle User Agent withing the limits of SDK as we don't expose it to SDK consumers

Reply:
I see the point now - it used to store useragent in file, not for API calls. LGTM